File: android/src/main/kotlin/com/appdynamics/appdynamics_mobilesdk/features/request_tracking/tracker_user_data/TrackerUserDataDate.kt
/*
 * Copyright (c) 2021. AppDynamics LLC and its affiliates.
 * All rights reserved.
 *
 */

package com.appdynamics.appdynamics_mobilesdk.features.request_tracking.tracker_user_data

import androidx.annotation.NonNull
import com.appdynamics.appdynamics_mobilesdk.AppDynamicsMobileSdkPlugin
import io.flutter.plugin.common.MethodChannel
import java.text.SimpleDateFormat
import java.util.*

@Suppress("SpellCheckingInspection")
fun AppDynamicsMobileSdkPlugin.setRequestTrackerUserDataDate(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val id = properties["id"] as String

    val tracker = requestTrackers[id] ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerUserDataDate() failed.",
            "Request tracker was not initialized or already reported."
        )
        return
    }

    val key = properties["key"] as? String ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerUserDataDate() failed.",
            "Please provide a valid string for `key`."
        )
        return
    }

    val value = properties["value"] as? String ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerUserDataDate() failed.",
            "Please provide a valid DateTime for `value`."
        )
        return
    }

    val date = SimpleDateFormat(
        dateFormat,
        Locale.US
    ).parse(value) ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerUserDataDate() failed.",
            "Please provide a valid DateTime for `value`."
        )
        return
    }

    tracker.withUserDataDate(key, date)
    result.success(null)
}

File: ios/Classes/Features/Request Tracking/Tracker User Data/TrackerUserDataDate.swift
import ADEUMInstrumentation
import Flutter

extension SwiftAppDynamicsAgentPlugin {
  func setRequestTrackerUserDataDate(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let id = properties["id"] as! String
    
    guard let key = properties["key"] as? String else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerUserDataDate() failed.", details: "Please provide a valid string for `key`.")
      result(error)
      return
    }
    
    guard let value = properties["value"] as? Double else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerUserDataDate() failed.", details: "Please provide a valid DateTime for `value`.")
      result(error)
      return
    }
      
    let date = Date(timeIntervalSince1970: value / 1000)
    
    guard let tracker = requestTrackers[id] else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerUserDataDate() failed.", details: "Request tracker was not initialized or already reported.")
      result(error)
      return
    }
    
    tracker.setUserDataDate(key, value: date)
    result(nil)
  }
}
